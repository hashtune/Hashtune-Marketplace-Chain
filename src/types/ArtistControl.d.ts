/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import {
  ethers,
  EventFilter,
  Signer,
  BigNumber,
  BigNumberish,
  PopulatedTransaction,
  BaseContract,
  ContractTransaction,
  Overrides,
  CallOverrides,
} from "ethers";
import { BytesLike } from "@ethersproject/bytes";
import { Listener, Provider } from "@ethersproject/providers";
import { FunctionFragment, EventFragment, Result } from "@ethersproject/abi";
import { TypedEventFilter, TypedEvent, TypedListener } from "./commons";

interface ArtistControlInterface extends ethers.utils.Interface {
  functions: {
    "approveArtist(address)": FunctionFragment;
    "approveArtistBatch(address[])": FunctionFragment;
    "approvedArtists(address)": FunctionFragment;
    "owner()": FunctionFragment;
    "renounceOwnership()": FunctionFragment;
    "revokeArtistApproval(address)": FunctionFragment;
    "revokeArtistBatchApproval(address[])": FunctionFragment;
    "transferOwnership(address)": FunctionFragment;
  };

  encodeFunctionData(
    functionFragment: "approveArtist",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "approveArtistBatch",
    values: [string[]]
  ): string;
  encodeFunctionData(
    functionFragment: "approvedArtists",
    values: [string]
  ): string;
  encodeFunctionData(functionFragment: "owner", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "renounceOwnership",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "revokeArtistApproval",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "revokeArtistBatchApproval",
    values: [string[]]
  ): string;
  encodeFunctionData(
    functionFragment: "transferOwnership",
    values: [string]
  ): string;

  decodeFunctionResult(
    functionFragment: "approveArtist",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "approveArtistBatch",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "approvedArtists",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "owner", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "renounceOwnership",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "revokeArtistApproval",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "revokeArtistBatchApproval",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "transferOwnership",
    data: BytesLike
  ): Result;

  events: {
    "ArtistApprovalRevoked(address)": EventFragment;
    "ArtistApproved(address)": EventFragment;
    "ArtistBatchApprovalRevoked(address[])": EventFragment;
    "ArtistBatchApproved(address[])": EventFragment;
    "OwnershipTransferred(address,address)": EventFragment;
  };

  getEvent(nameOrSignatureOrTopic: "ArtistApprovalRevoked"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "ArtistApproved"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "ArtistBatchApprovalRevoked"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "ArtistBatchApproved"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "OwnershipTransferred"): EventFragment;
}

export class ArtistControl extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  listeners<EventArgsArray extends Array<any>, EventArgsObject>(
    eventFilter?: TypedEventFilter<EventArgsArray, EventArgsObject>
  ): Array<TypedListener<EventArgsArray, EventArgsObject>>;
  off<EventArgsArray extends Array<any>, EventArgsObject>(
    eventFilter: TypedEventFilter<EventArgsArray, EventArgsObject>,
    listener: TypedListener<EventArgsArray, EventArgsObject>
  ): this;
  on<EventArgsArray extends Array<any>, EventArgsObject>(
    eventFilter: TypedEventFilter<EventArgsArray, EventArgsObject>,
    listener: TypedListener<EventArgsArray, EventArgsObject>
  ): this;
  once<EventArgsArray extends Array<any>, EventArgsObject>(
    eventFilter: TypedEventFilter<EventArgsArray, EventArgsObject>,
    listener: TypedListener<EventArgsArray, EventArgsObject>
  ): this;
  removeListener<EventArgsArray extends Array<any>, EventArgsObject>(
    eventFilter: TypedEventFilter<EventArgsArray, EventArgsObject>,
    listener: TypedListener<EventArgsArray, EventArgsObject>
  ): this;
  removeAllListeners<EventArgsArray extends Array<any>, EventArgsObject>(
    eventFilter: TypedEventFilter<EventArgsArray, EventArgsObject>
  ): this;

  listeners(eventName?: string): Array<Listener>;
  off(eventName: string, listener: Listener): this;
  on(eventName: string, listener: Listener): this;
  once(eventName: string, listener: Listener): this;
  removeListener(eventName: string, listener: Listener): this;
  removeAllListeners(eventName?: string): this;

  queryFilter<EventArgsArray extends Array<any>, EventArgsObject>(
    event: TypedEventFilter<EventArgsArray, EventArgsObject>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEvent<EventArgsArray & EventArgsObject>>>;

  interface: ArtistControlInterface;

  functions: {
    approveArtist(
      artist: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    approveArtistBatch(
      artists: string[],
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    approvedArtists(
      arg0: string,
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    owner(overrides?: CallOverrides): Promise<[string]>;

    renounceOwnership(
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    revokeArtistApproval(
      artist: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    revokeArtistBatchApproval(
      artists: string[],
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    transferOwnership(
      newOwner: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;
  };

  approveArtist(
    artist: string,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  approveArtistBatch(
    artists: string[],
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  approvedArtists(arg0: string, overrides?: CallOverrides): Promise<boolean>;

  owner(overrides?: CallOverrides): Promise<string>;

  renounceOwnership(
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  revokeArtistApproval(
    artist: string,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  revokeArtistBatchApproval(
    artists: string[],
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  transferOwnership(
    newOwner: string,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    approveArtist(artist: string, overrides?: CallOverrides): Promise<void>;

    approveArtistBatch(
      artists: string[],
      overrides?: CallOverrides
    ): Promise<void>;

    approvedArtists(arg0: string, overrides?: CallOverrides): Promise<boolean>;

    owner(overrides?: CallOverrides): Promise<string>;

    renounceOwnership(overrides?: CallOverrides): Promise<void>;

    revokeArtistApproval(
      artist: string,
      overrides?: CallOverrides
    ): Promise<void>;

    revokeArtistBatchApproval(
      artists: string[],
      overrides?: CallOverrides
    ): Promise<void>;

    transferOwnership(
      newOwner: string,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {
    ArtistApprovalRevoked(
      artist?: null
    ): TypedEventFilter<[string], { artist: string }>;

    ArtistApproved(
      artist?: null
    ): TypedEventFilter<[string], { artist: string }>;

    ArtistBatchApprovalRevoked(
      artists?: null
    ): TypedEventFilter<[string[]], { artists: string[] }>;

    ArtistBatchApproved(
      artists?: null
    ): TypedEventFilter<[string[]], { artists: string[] }>;

    OwnershipTransferred(
      previousOwner?: string | null,
      newOwner?: string | null
    ): TypedEventFilter<
      [string, string],
      { previousOwner: string; newOwner: string }
    >;
  };

  estimateGas: {
    approveArtist(
      artist: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    approveArtistBatch(
      artists: string[],
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    approvedArtists(
      arg0: string,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    owner(overrides?: CallOverrides): Promise<BigNumber>;

    renounceOwnership(
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    revokeArtistApproval(
      artist: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    revokeArtistBatchApproval(
      artists: string[],
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    transferOwnership(
      newOwner: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    approveArtist(
      artist: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    approveArtistBatch(
      artists: string[],
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    approvedArtists(
      arg0: string,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    owner(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    renounceOwnership(
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    revokeArtistApproval(
      artist: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    revokeArtistBatchApproval(
      artists: string[],
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    transferOwnership(
      newOwner: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;
  };
}
