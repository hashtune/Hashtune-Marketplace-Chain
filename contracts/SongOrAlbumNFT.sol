//SPDX-License-Identifier: Unlicense
pragma solidity ^0.8.0;

import "hardhat/console.sol";
import "@openzeppelin/contracts/access/AccessControl.sol";
import "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
import "@openzeppelin/contracts/access/Ownable.sol";
import "./DataModel.sol";
// override key word means it is overriding a virtual function it inherited from
// public - all can access
// external - Cannot be accessed internally, only externally
// internal - only this contract and contracts deriving from it can access
// private - can be accessed only from this contract
// TODO add event emissions
contract SongOrAlbumNFT is ERC1155, Ownable, AccessControl {

    uint256 totalArts;
    address hashtuneAddress;
    uint256 hashtuneShare = 2; //represents 2%
    uint256 artistRoyalty = 2;
    
    //TODO: refactor to use different structure in order to save some storage
    mapping(uint256 => DataModel.ArtInfo) public arts;
    mapping(uint256 => uint256) public _prices;
    mapping(uint256 => address[]) private _tokenCreators;
    mapping(uint256 => address) private _currentOwners;
    mapping(uint256 => bool) public _listings;
    mapping(uint256 => mapping(uint256 => DataModel.AuctionInfo)) public bids;
    mapping(address => mapping(uint256 => mapping(uint256 => uint256))) public bidMoneyPool;
    mapping(uint256 => uint256) public totalAuctions; // alternative to mapping array of struct 

    // Custom Events
    event NewURI(address setBy, string newAddress);
    event TokenCreated(address by, uint256 tokenId, address[] creators);
    event TokenPurchased(address by, uint256 tokenId);
    event PayoutOccurred(address to, uint256 amount);
    event NewPrice(address setBy, uint256 newPrice, uint256 tokenId);
    event NewBid(address by, uint256 tokenId, uint256 amount);
    event NewAuction(uint256 tokenId, uint256 auctionNum, uint256 reservePrice, uint256 endTime);
    event EndAuction(uint256 tokenId, uint256 auctionNum, address newOwner, uint256 soldFor);
    event WithdrawMoney(address receiver, uint256 withdrawnAmount);
    
    constructor (string memory uri_, uint256 _hashtuneShare, uint256 _artistRoyalty) ERC1155(uri_) {
        console.log("Deploying a Song or Album Contract with uri:", uri_);
        hashtuneAddress = msg.sender;
        hashtuneShare = _hashtuneShare; //adding share value at the time of deployment
        artistRoyalty = _artistRoyalty;
    }

    function supportsInterface(bytes4 interfaceId) public view virtual override(ERC1155, AccessControl) returns (bool) {
        return super.supportsInterface(interfaceId);
    }
    
    // Example of such a URI: https://token-cdn-domain/{id}.json would be replaced with https://token-cdn-domain/000000000000000000000000000000000000000000000000000000000004cce0.json if the client is referring to token ID 314592/0x4CCE0.
    function setURI(string memory newuri) public onlyOwner {
        emit NewURI(msg.sender, newuri);
        _setURI(newuri);
    }

    // One URI for all tokens
    function showURI(uint256 id) public view returns (string memory) {
        return uri(id);
    }

    function showSalePriceFor(uint256 id) public view returns (uint256) {
        return _prices[id];
    }

    function create(
        address[] memory creators,
        uint256[] memory creatorsShare,
        DataModel.ArtStatus status,
        bytes memory data,
        uint256 salePrice,
        bytes32 digest,
        uint8 hashFunction,
        uint8 size
        ) public {
        
        require(creators.length == creatorsShare.length, "all creators must have shares defined");
        
        arts[totalArts++] = DataModel.ArtInfo(
            status,
            msg.sender,
            creators,
            creatorsShare,
            salePrice,
            DataModel.MultiHash(
                digest,
                hashFunction,
                size));
        
        _mint(msg.sender, totalArts, 1, data);
        emit TokenCreated(msg.sender, totalArts, creators);
    }

    function setApprovalToBuy(address toApprove, uint256 tokenId) public {
        // Emits setApprovalForAll
        require(_currentOwners[tokenId] == msg.sender, "cannot set approval if not token owner");
        return setApprovalForAll(toApprove, true);
    }
  
    function buy(uint256 tokenId) public payable {
        // Emits a TransferSingle 
        emit TokenPurchased(msg.sender, tokenId);
        uint256 price = showSalePriceFor(tokenId);
        require(msg.value == price, "incorrect amount sent");
        //require(_listings[tokenId] == true, "art piece not for sale"); only makes sense if theres one
        require(msg.sender != address(0) && msg.sender != address(this), "cannot buy your own token");

        // Royalty
        // TODO customize the royalty for each feature
        // Assuming ether and not wei (10^18 ether)
        // Need to restrict the number of features
        uint256 creatorRoyalty = (msg.value * 2) / 100;
        uint256 _tokenCreatorsLength = _tokenCreators[tokenId].length;
        for (uint256 i=0; i<_tokenCreatorsLength; i++) {
            sendTo(payable(_tokenCreators[tokenId][i]), creatorRoyalty);
        }
        
        // Hashtune Cut
        uint256 platformCut = (msg.value * 2) / 100;
        sendTo(payable(hashtuneAddress), platformCut);

        // Current Owner
        // On the first transaction the current owner is the first creator
        // and thus does not receive a royalty
        uint256 amount = msg.value - (creatorRoyalty * _tokenCreatorsLength) - platformCut;
        sendTo(payable(_currentOwners[tokenId]), amount);
        bytes memory data;

        // Transfer token
        // Need to actually transfer the contract if using 721Upgradeable (?)
        safeTransferFrom(_currentOwners[tokenId], msg.sender, tokenId, 1, data);
        _currentOwners[tokenId] = msg.sender;
    }

    function sendTo(address payable receiver, uint256 _amount) private {
        // Not sure if this emits a default transfer event?
        emit PayoutOccurred(receiver, _amount);
        require(receiver != address(0) && receiver != address(this), "receiver is contract address owner");
        require(_amount > 0 && _amount <= address(this).balance, "amount is less than contract balance");
        receiver.transfer(_amount);
    }
    
    // set listed? This will cost gas to set but prevents a sale from happening without current owners consent?
    function getCurrentOwner(uint256 tokenId) view public returns (address) {
        return _currentOwners[tokenId];
    }

    // Set current price and buy could have a race condition, make sure you 
    // TODO: disable purchase while updating prices by pausing contract first then unpausing
    function setCurrentPrice(uint256 newPrice, uint256 tokenId) public  {
        emit NewPrice(msg.sender, newPrice, tokenId);
        require(msg.sender == _currentOwners[tokenId], "cannot set the price for a token you don't currently own");
        require(newPrice > 0, "cannot set the new price of the token to zero");
        _prices[tokenId] = newPrice;
    }

    //Start the Auction on limeted time based model or without time contraint
    // TODO: implementing the limited time based auction model
    /** 
    * If the reserve price set is 0, there is no duration for this auction. 
    * If the reserve price is set > 0, there is a duration for this auction.
     */
    function startAuction(uint256 tokenId, uint256 reservePrice) public {
        address currentOwner = _currentOwners[tokenId];
        uint256 previousAuctions = totalAuctions[tokenId];
        bool wasFinalized = bids[tokenId][previousAuctions].isFinalized;
        if(previousAuctions > 0) {
            require(wasFinalized, "previous auction is still on going.");
        }
        uint256 endTime = block.timestamp + (24*60*60);
        require(currentOwner == msg.sender, "can't start the auction of NFT you don't own");
        uint256 newAuctionNum = totalAuctions[tokenId] += 1;
        if(reservePrice > 0) {
            require(endTime > block.timestamp, "auction endtime should be set to future");
            bids[tokenId][newAuctionNum].endTime = endTime;
        } else {
            bids[tokenId][newAuctionNum].endTime = 0;
        }
        bids[tokenId][newAuctionNum].reservePrice = reservePrice;
        bids[tokenId][newAuctionNum].currentHigh = reservePrice;
        bids[tokenId][newAuctionNum].isFinalized = false;
        emit NewAuction(tokenId, newAuctionNum, bids[tokenId][newAuctionNum].reservePrice, endTime);
    }

    //TODO: implement safe check for whether NFT is up for auction
    function placeBid(uint256 tokenId) public payable {
        uint256 currentAuctionNum = totalAuctions[tokenId];
        uint256 newBidSum = msg.value + bidMoneyPool[msg.sender][tokenId][currentAuctionNum];
        uint256 endTime = bids[tokenId][currentAuctionNum].endTime;
        bool isEnded = bids[tokenId][currentAuctionNum].isFinalized;
        require(!isEnded, "auction is closed for this NFT");
        // Auction numbers start at 1 because token Numbers start at 1
        require(currentAuctionNum > 0, "no ongoing auction for this NFT");
        if(endTime == 0 || block.timestamp < endTime) {
            require(msg.value > 0, "bid amount should be greator than 0");
            require(newBidSum > bids[tokenId][currentAuctionNum].reservePrice, "bid amount should be greator than the reservePrice");
            require(newBidSum > bids[tokenId][currentAuctionNum].currentHigh, "bid amount should be greator than the current highest");
            bidMoneyPool[msg.sender][tokenId][currentAuctionNum] += msg.value;
            bids[tokenId][currentAuctionNum].currentHigh = newBidSum;
            bids[tokenId][currentAuctionNum].currentHighBider = msg.sender;
        } else {
            revert("auction is closed for this NFT");
        }
        emit NewBid(msg.sender, tokenId, msg.value);
    }

    //TODO: implement safe check for whether NFT is up for auction, spliting the royalties, transfering the money
    function endAuction(uint256 tokenId) public {
        uint256 currentAuctionNum = totalAuctions[tokenId];
        require(_currentOwners[tokenId] == msg.sender, "only owner of the NFT can end the auction");
         bytes memory data;
        if(bids[tokenId][currentAuctionNum].endTime == 0) {
            safeTransferFrom(_currentOwners[tokenId], bids[tokenId][currentAuctionNum].currentHighBider, tokenId, 1, data);
        } else {
            require(bids[tokenId][currentAuctionNum].endTime < block.timestamp, "can`t end ongoing time based auction");
            safeTransferFrom(_currentOwners[tokenId], bids[tokenId][currentAuctionNum].currentHighBider, tokenId, 1, data);
        }
        bids[tokenId][currentAuctionNum].isFinalized = true;
        emit EndAuction(tokenId, currentAuctionNum, bids[tokenId][currentAuctionNum].currentHighBider, bids[tokenId][currentAuctionNum].currentHigh);
    }

    //lets you withdraw your bid money when the auction is ended.
    //TODO: implement withdraw all the previous money in case the current auction is ongoing
    function withdrawBidMoney(uint tokenId) public {
        uint256 currentAuctionNum = totalAuctions[tokenId];
        uint256 currentAuctionBalance = bidMoneyPool[msg.sender][tokenId][currentAuctionNum];
        bool isFinalized = bids[tokenId][currentAuctionNum].isFinalized;
        uint256 balance;
        if(currentAuctionBalance == 0) {
            revert("you don't have any money in the bidding pool");
        } else {
            require(isFinalized, "can't withdraw money until the auction has ended");
            balance = bidMoneyPoolCalculator(tokenId, currentAuctionNum, msg.sender);
            if(balance > 0) {
                if(!payable(msg.sender).send(balance)) {
                    revert("withdrawal unsuccessful");
                }
            } else {
                revert("you don't have any money in the bidding pool");
            }
        }
        emit WithdrawMoney(msg.sender, balance);
    }

    //helper function to save storage
    function bidMoneyPoolCalculator(uint256 tokenId, uint256 toAuctionNum, address bider) private returns (uint256) {
        require(toAuctionNum > 0, "no previous auctions happened for this NFT");
        uint256 balance;
        for(uint256 i=1; i <= toAuctionNum; i++) {
                balance += bidMoneyPool[bider][tokenId][i];
                bidMoneyPool[bider][tokenId][i] = 0;
        }
        return balance;
    }
}