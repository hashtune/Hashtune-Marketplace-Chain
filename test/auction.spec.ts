import "@nomiclabs/hardhat-ethers"; // Adds ethers property object to hardhat run time environment
import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
import chai from "chai";
import hre from "hardhat";
import { ethers } from "hardhat";
import { jestSnapshotPlugin } from "mocha-chai-jest-snapshot";

import { SongOrAlbumNFT } from "../src/types/SongOrAlbumNFT";

import { connectAsUser } from "./utils";
chai.use(jestSnapshotPlugin());
const { expect } = chai;

export interface Context {
  soa: SongOrAlbumNFT;
  users: { [key: string]: SignerWithAddress };
}

describe("auction", async () => {
  let context: Context;
  const tokenId = 1;
  // This is super cheap, this is a 17 USD token if 1 Ether is worth 3.4K USD
  const tokenPrice = ethers.utils.parseEther("0.005");
  const newTokenPrice = ethers.utils.parseEther("0.006");
  const highestTokenPrice = ethers.utils.parseEther("0.008");
  const lowerTokenPrice = ethers.utils.parseEther("0.004");
  beforeEach(async () => {
    const [one, two, three, four, five, six] = await ethers.getSigners();
    const SongOrAlbum = await hre.ethers.getContractFactory("SongOrAlbumNFT");
    const SOA: SongOrAlbumNFT = (await SongOrAlbum.deploy(
      "http://blank"
    )) as SongOrAlbumNFT;
    // Deploy the contract
    await SOA.deployed();
    context = {
      soa: SOA,
      users: {
        one, // contract owner and token creator
        two,
        three, // token feature
        four,
        five,
        six,
      },
    };
    // Create a single token
    const result = await context.soa.create(
      [context.users.one.address, context.users.three.address],
      tokenId,
      [],
      tokenPrice
    );
    if (!result.hash) {
      throw new Error("Problem setting up tests");
    }
    const eventFilter = context.soa.filters.TransferSingle();
    return SOA;
  });

  describe("Auction with reserve price", () => {
    it("can create an auction for a token I own with a reserve price", async () => {
      const res = await context.soa.startAuction(1, tokenPrice);
      expect(res.from).toMatchSnapshot();
    });
    it("cannot create an auction for a token I do not own", async () => {
      const asTwo = connectAsUser(context.users.two, context);
      try {
        await asTwo.startAuction(1, tokenPrice);
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
    it("cannot create an auction for a token that already has an ongoing auction", async () => {
      await context.soa.startAuction(1, tokenPrice);
      try {
        await context.soa.startAuction(1, tokenPrice);
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
    it("can bid on an auction if my price is higher than the reserve price", async () => {
      await context.soa.startAuction(1, tokenPrice);
      const asTwo = connectAsUser(context.users.two, context);
      const res = await asTwo.placeBid(1, { value: newTokenPrice });
      expect(res.from).toMatchSnapshot();
    });
    it("cannot bid on an auction if my price is lower than the reserve price", async () => {
      await context.soa.startAuction(1, tokenPrice);
      const asTwo = connectAsUser(context.users.two, context);
      try {
        const res = await asTwo.placeBid(1, { value: lowerTokenPrice });
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
    it("cannot bid on an auction if my price is lower than the current high", async () => {
      await context.soa.startAuction(1, tokenPrice);
      const asTwo = connectAsUser(context.users.two, context);
      const asThree = connectAsUser(context.users.three, context);
      await asTwo.placeBid(1, { value: newTokenPrice });
      try {
        await asThree.placeBid(1, { value: lowerTokenPrice });
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
    it("can bid on an auction if my price is higher than the current high", async () => {
      await context.soa.startAuction(1, tokenPrice);
      const asTwo = connectAsUser(context.users.two, context);
      await asTwo.placeBid(1, { value: newTokenPrice });
      const asThree = connectAsUser(context.users.three, context);
      const res = await asThree.placeBid(1, { value: highestTokenPrice });
      expect(res.from).toMatchSnapshot();
    });
    it("can bid a lower reserve price and currenthigh if I already have a bid, it should be cumulative", async () => {
      await context.soa.startAuction(1, tokenPrice);
      const asTwo = connectAsUser(context.users.two, context);
      await asTwo.placeBid(1, { value: newTokenPrice });
      const res = await asTwo.placeBid(1, { value: lowerTokenPrice });
      expect(res.from).toMatchSnapshot();
    });
    it.skip("cannot bid on an auction once it has ended", async () => {
      // You cannot close an auction that has no bids.
      await context.soa.startAuction(1, tokenPrice);
      const asTwo = connectAsUser(context.users.two, context);
      await asTwo.placeBid(1, { value: newTokenPrice });
      await context.soa.endAuction(1);
      try {
        await asTwo.placeBid(1, { value: newTokenPrice });
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
    it("cannot end an ongoing auction before the end time", async () => {
      // You cannot close an auction that has no bids.
      await context.soa.startAuction(1, tokenPrice);
      const asTwo = connectAsUser(context.users.two, context);
      await asTwo.placeBid(1, { value: newTokenPrice });
      try {
        await context.soa.endAuction(1);
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
    it.skip("can end an auction that I created", () => {});
    it("cannot end an auction I did not create", async () => {
      await context.soa.startAuction(1, tokenPrice);
      const asTwo = connectAsUser(context.users.two, context);
      try {
        await asTwo.endAuction(1);
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
    it.skip("can receive the token if I am the last highest bidder", async () => {
      await context.soa.startAuction(1, tokenPrice);
      const asTwo = connectAsUser(context.users.two, context);
      await asTwo.placeBid(1, { value: newTokenPrice });
      const asThree = connectAsUser(context.users.three, context);
      await asThree.placeBid(1, { value: highestTokenPrice });
      //End auction after 24hours and check new Owner and funds
      // expect(res.from).toMatchSnapshot();
    });
    it("cannot withdraw my funds out if the auction is ongoing", async () => {
      await context.soa.startAuction(1, tokenPrice);
      const asTwo = connectAsUser(context.users.two, context);
      await asTwo.placeBid(1, { value: newTokenPrice });
      try {
        await asTwo.withdrawBidMoney(1);
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
    it("cannot withdraw my funds out if I did not put money in", async () => {
      await context.soa.startAuction(1, tokenPrice);
      const asTwo = connectAsUser(context.users.two, context);
      const asThree = connectAsUser(context.users.three, context);
      await asTwo.placeBid(1, { value: newTokenPrice });
      try {
        await asThree.withdrawBidMoney(1);
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
    it.skip("can withdraw my funds out after the auction ends if I did not win", () => {});
    it.skip("cannot withdraw my funds out after the auction ends if I won", () => {});
    it.skip("can receive payment after the auction ends if I am the token owner", () => {});
    it.skip("can receive royalties after the auction ends if I am the token creator", () => {});
  });

  describe("Auction with no reserve price", () => {
    it("can create an auction for a token I own without a reserve price", () => {});
    it("cannot create an auction for a token I do not own", () => {});
    it("cannot create an auction for a token that already has an ongoing auction", async () => {});
    it("can bid on an auction if my price is higher than the current high", () => {});
    it("cannot bid on an auction if my price is lower than the current high", () => {});
    it("cannot bid on an auction once it has ended", () => {});
    it("can bid on an auction multiple times and have money pooled", () => {});
    it("can end an auction that I created", () => {});
    it("cannot end an auction I did not create", () => {});
    it("can receive the token if I am the last highest bidder when the auction ends", () => {});
    it("can withdraw my funds out after the auction ends if I did not win", () => {});
    it("cannot withdraw my funds out after the auction ends if I won", () => {});
    it("can receive payment after the auction ends if I am the token owner", () => {});
    it("can receive royalties after the auction ends if I am the token creator", () => {});
  });
  describe.skip("events", () => {});
  describe.skip("serially incrementing token Ids", () => {});
});
