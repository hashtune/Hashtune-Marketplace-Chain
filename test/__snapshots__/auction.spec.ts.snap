// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auction Auction with reserve price can bid a lower reserve price and currenthigh if I already have a bid, it should be cumulative 1`] = `"0x70997970C51812dc3A010C7d01b50e0d17dc79C8"`;

exports[`auction Auction with reserve price can bid on an auction if my price is higher than the current high 1`] = `"0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC"`;

exports[`auction Auction with reserve price can bid on an auction if my price is higher than the reserve price 1`] = `"0x70997970C51812dc3A010C7d01b50e0d17dc79C8"`;

exports[`auction Auction with reserve price can create an auction for a token I own with a reserve price 1`] = `"0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"`;

exports[`auction Auction with reserve price cannot bid on an auction if my price is lower than the current high 1`] = `[ProviderError: Error: VM Exception while processing transaction: reverted with reason string 'bid amount should be greator than the reservePrice']`;

exports[`auction Auction with reserve price cannot bid on an auction if my price is lower than the reserve price 1`] = `[ProviderError: Error: VM Exception while processing transaction: reverted with reason string 'bid amount should be greator than the reservePrice']`;

exports[`auction Auction with reserve price cannot create an auction for a token I do not own 1`] = `[ProviderError: Error: VM Exception while processing transaction: reverted with reason string 'can't start the auction of NFT you don't own']`;

exports[`auction Auction with reserve price cannot create an auction for a token that already has an ongoing auction 1`] = `[ProviderError: Error: VM Exception while processing transaction: reverted with reason string 'previous auction is still on going.']`;

exports[`auction Auction with reserve price cannot end an auction I did not create 1`] = `[ProviderError: Error: VM Exception while processing transaction: reverted with reason string 'only owner of the NFT can end the auction']`;

exports[`auction Auction with reserve price cannot end an ongoing auction before the end time 1`] = `[ProviderError: Error: VM Exception while processing transaction: reverted with reason string 'can\`t end ongoing time based auction']`;

exports[`auction Auction with reserve price cannot withdraw my funds out if I did not put money in 1`] = `[ProviderError: Error: VM Exception while processing transaction: reverted with reason string 'you don't have any money in the bidding pool']`;

exports[`auction Auction with reserve price cannot withdraw my funds out if the auction is ongoing 1`] = `[ProviderError: Error: VM Exception while processing transaction: reverted with reason string 'can't withdraw money until the auction has ended']`;
